### ISSUE TITLE
FEATURE: Switch from flags to language names

### ISSUE THREAD

**Issue by twatzl** [pacollins/hugo-future-imperfect-slim#61]
# Bug Report

## Problem description
I am from Austria. Although we speak German here our language and culture is different from Germany in a few aspects. 

Our official locale as far as i know is de-AT, but when I use that for the translation it tries to load the flag `flag-icon-de-at` which does not exist.

## Solution
A configuration possibility to configure the flag icon independently could solve the issue.

i.e.
```
flag: at
```

## Workaround
I just use 'at' now as locale. Now the flag works, but I am not sure what else I broke in the process of doing so.

**Comment by twatzl** [comment]
Sorry for the confusing label. Initially thought this would become a feature request.

**Comment by pacollins** [comment]
For English, [I created a merged flag](https://github.com/pacollins/hugo-future-imperfect-slim/tree/master/static/flags) under `en`.

In my opinion, the Language flags are more for a visual representation of the main language. I have seen merged USA/GB flags, but rarely merged other flags. If we feel like this is a big deal, it would be worth exploring, but ultimate, the languages are "English" and "German".

That being said, I am open to having a conversation about it.

**Comment by twatzl** [comment]
Does using 'at' as locale have any negative impact for site users? For example because their browser does not choose the right language automatically.
If not I am fine with this workaround.

Although it seems a picky I do not really feel comfortable with using the German flag instead. After all it is a personal blog and I am not from Germany. This is especially true since 'Austrian German' can be quite different from German German.

For creating a merged flag you would need also to include Switzerland, since they speak German there as well, so it would get quite crowded in there ;)

Another workaround could also be to just add an option to disable the flags and have a language dropdown or something. But as I said, if my workaround has no negative impact then I am fine with that.

**Comment by pacollins** [comment]
After thinking a little further, the _easiest_ solution would be to add different locales to the `style.scss` here:

https://github.com/pacollins/hugo-future-imperfect-slim/blob/64535832649fab584fb81d6391ad70e25e512515/static/css/main.scss#L761-L773

That being said, if you are not translating your blog, you should just turn multilingual off and remove the flags. Either by disabling languages:

https://github.com/pacollins/hugo-future-imperfect-slim/blob/64535832649fab584fb81d6391ad70e25e512515/exampleSite/config.toml#L10

or completely removing the from the `config.toml`:

https://github.com/pacollins/hugo-future-imperfect-slim/blob/64535832649fab584fb81d6391ad70e25e512515/exampleSite/config.toml#L139-L255

**Comment by VincentTam** [comment]
> Although it seems a picky I do not really feel comfortable with using the German flag instead. After all it is a personal blog and I am not from Germany. This is especially true since 'Austrian German' can be quite different from German German.
r
@twatzl Born and raised in :hong_kong:, a former :gb: British colony, I share similar feelings.  Three months ago, a GitLab issue to merge Traditional Chinese locales :hong_kong: zh-HK and :taiwan: zh-TW caused a storm in the teacup: https://gitlab.com/gitlab-org/rgitlab-ce/issues/60768.

It would be great if you could offer us some translations, even a few locale strings.  They can be put into a pull request, which will be merged upon completion.  The PR for Spanish translation #52 is an example.

> For creating a merged flag you would need also to include Switzerland, since they speak German there as well, so it would get quite crowded in there ;)

:smiley: Good point!  As a native :hong_kong: Cantonese speaker, I can't agree with you more.  The case for Chinese is even more illustrating: :macau: & :hong_kong:, :taiwan:, :singapore: & :malaysia:.

Perhaps we can learn how to handle different locales from Minimal Mistakes: https://github.com/mmistakes/minimal-mistakes/blob/43b5e5f2ef115f3467fd4fa7256ca38953d5e993/_data/ui-text.yml#L518-L529.

**Comment by twatzl** [comment]
@pacollins what about defining the flag string in the translation file? Because flag-icon-at already exists and works when you just name the translation file at.toml instead of de-AT.toml. This would basically resolve all those corner cases at once.

And in the *.toml file you would just define the flag with
`flag: flag-icon-at` and default behavior could remain unchanged if that property is not set.

And yes @pacollins I am using multilanguage feature. I write in German and English. This was the main reason why I switched from Hexo to Hugo and to this theme.

@VincentTam How would you recommend doing the translations? Should I just provide the same files for Austrian and German. After all the translations would probably be identical except for the flag. (The local varieties of the language are mostly different in the words for certain foods)

Should I just add 2 files with almost identical content?

**Comment by VincentTam** [comment]
I'm not (yet?) familiar with the working of the flag feature, so I'll leave @<span>pacollins</span> to respond.



> @VincentTam How would you recommend doing the translations? Should I just provide the same files for Austrian and German. After all the translations would probably be identical except for the flag. (The local varieties of the language are mostly different in the words for certain foods)

1. Fork this repo (online on :octocat: GitHub ).
2. Clone any one of the locale files `i18n/*.toml` and name the clones as `i18n/de-AT.toml` and `i18n/de-DE.toml` (or more)
3. Edit the double quoted string in `other = "..."` under each field.
4. Review and commit into a new branch, say `DE`.  You can choose whatever name for the new branch.
5. Create pull request and celebrate :tada:.

> Should I just add 2 files with almost identical content?

Why not?

### TL;DR
In this case, developers would normally use one single `de.toml` file for them.  By given the special situtation of this theme (languages got represented by regional/national flags, which is :thinking:—what flag should I use for Cantonese, my native language?  :hong_kong: ?  But Macau people :macau:, ... also speak it, with some little differences of course.), I'll be open to the possible duplication of `de-AT`, `de-DE`, `de-CH` and `de-BE`, you name it.  Anyways, provided that they are *identical*, that'll be stored in the Git repo as the *same* blob.  Having a few more entries in the Git index can't hurt, since everything is compressed.

Having language + regional codes in `i18n` files shows our prospective users that Hugo's i18n supports regional (differences in) languages.  IMO, that's a good sign.  In case of new fields in the future, these *identical* files might evolve and diverge, just like our languages and cultures.

**Comment by twatzl** [comment]
Ok. Fine. But I am going to wait until we have a proper approach for fixing the flag. I don't want to create a PR with a workaround. :D

**Comment by pacollins** [comment]
@twatzl Wouldn't doing `at.toml` provide the wrong locale tag? Shouldn't the locale tag be `de-at`? That is where I got lost.

**Comment by pacollins** [comment]
After experimenting, Hugo does not recognize `de-at` or `at` as a `defaultContentLanguage`.

```
Error: site config value "de-AT" for defaultContentLanguage does not match any language definition
Error: site config value "at" for defaultContentLanguage does not match any language definition
```
So this might be an issue to push further up the chain.

**Comment by pacollins** [comment]
So after looking further once more the line below _must_ be a

**Comment by pacollins** [comment]
The work around that you are using seems to be appropriate based on my usage. For the line below, Hugo requires it be an [ISO 639-1 Code](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) which are 2 letter codes (eg. `en`, `fr`, `de`). Your options would be to leave this blank or use an applicable code.

https://github.com/pacollins/hugo-future-imperfect-slim/blob/afbf3ccbc8889f3dbacbade26c8bf4bbb18d2e64/exampleSite/config.toml#L2

Then for the below lines, you can actually define it however you want. For example, you can replace `fr` with `at` at all locations, and the theme (and hugo) will function properly. Ironically, `lang="at"` validates while `lang="us"` does not.

https://github.com/pacollins/hugo-future-imperfect-slim/blob/afbf3ccbc8889f3dbacbade26c8bf4bbb18d2e64/exampleSite/config.toml#L147-L152

That being said, I am going to close this issue because the current solution does not require changing the theme. It might be worthwhile posting an issue on the Hugo repo though about them expanding applicable language/locale codes.

Feel free to continue the discussion here though, if needed.

**Comment by VincentTam** [comment]
> The work around that you are using seems to be appropriate based on my usage. For the line below, Hugo requires it be an [ISO 639-1 Code](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes) which are 2 letter codes (eg. `en`, `fr`, `de`). Your options would be to leave this blank or use an applicable code.

# :man_facepalming: 
Cantonese isn't found in the linked list, but in ISO 639-3.  Anyways, just a comment.  Most Cantonese speakers can read from either :taiwan: zh-TW or :cn: zh-CN.

**Comment by martignoni** [comment]
You could also [stop using flags to represent languages](http://www.flagsarenotlanguages.com/blog/why-flags-do-not-represent-language/).

**Comment by VincentTam** [comment]
Another suggestion for keeping flags copied from https://discourse.gohugo.io/t/i18n-support-for-language-country-code/20303/4?u=vincenttam:

Try:

> ```toml
> languageCode = "de-at"
> ```
>
> For a multilingual website you can use this:
> ```toml
> defaultContentLanguage = "de"
> [languages]
>   [languages.de]
>     weight = 1
>     languageCode = "de-at"
>   [languages.en]
>     # …
> ```
> In your `baseof.html`, you may try
> ```html
> <html lang="{{ .Site.LanguageCode }}">
> ```

**Comment by pacollins** [comment]
@VincentTam Let's explore that and see how it looks.

@martignoni I understand the sentiment, but its more about easy of access. Across the internet this is a common feature people are used to using. That being said, I am open to a suggestion as a replacement as long as it isn't an eyesore.

**Comment by martignoni** [comment]
@pacollins Re: flags do not represent languages. E.g. my country (Switzerland) has four official languages, but only one official flag. So it's maybe more than a sentiment. Anyway, you're right that the practice to use flag for languages is relatively spread in the Internet. So why not take this opportunity to make the Internet better? :-)

Some best practices here: http://www.flagsarenotlanguages.com/blog/best-practice-for-presenting-languages/.

And here's how I'm doing it in one of my (Hugo based) website.
![lang](https://user-images.githubusercontent.com/421392/63266210-b5ef9500-c28f-11e9-910d-8135e50490ea.png)

**Comment by pacollins** [comment]
Wouldn't that ultimately lead to the same issue as the original report? Or
would you recommend doing Deutsch-AT (or similar)?

**Comment by martignoni** [comment]
I would imagine a menu similar to this:
- English (UK)
- English (US)
- Deutsch (AT)
- Deutsch (DE)
- Deutsch (CH)
- Français (FR)
- Français (CA)
- Français (CH)

And possibly, if there's only one language variation, omit the info in the parenthesis.

**Comment by twatzl** [comment]
@pacollins I am using at.toml. Probably I did not experience the issues you describe, because I am still using English as the main language for my blog and German just as secondary.

I think a menu as described by @martignoni would look nice. I don't see why the language choosing menu has to take up the entire right side anyway.

Regarding the issue of differentiating between languages in the written menu: I don't think this problem exists. The flag problem only exists, because I am Austrian and not German, but the language is still called German and nobody will translate it in all 3 languages. I have no idea how that is with other languages, but I think it is still save to say that either the languages are similar enough to be called the same and don't need translation or are different enough to have their own names.

**Comment by pacollins** [comment]
So you'd be fine with ignoring locales and just saying the overarching
language? That should be pretty easy.

It's a sliding side bar for stylistic consistency. We don't currently have
any dropdown menus, so it wouldn't make sense.

**Comment by VincentTam** [comment]
> We don't currently have any dropdown menus, so it wouldn't make sense.

After fixing #46, that would be possible with additional CSS settings, like [my blog](https://vincenttam.gitlab.io) using Beautiful Hugo.

**Comment by VincentTam** [comment]
@twatzl Luckily, someone has already translated most of the UI strings for us:
https://github.com/marcelkraus/twietesla/blob/978fe99689adc1c38ce064f474ca554dfbcdfc1f/i18n/de.toml.

I've put this into #75.  As I don't know German, I'm open to changes.

> Regarding the issue of differentiating between languages in the written menu: I don't think this problem exists. The flag problem only exists, because I am Austrian and not German, but the language is still called German and nobody will translate it in all 3 languages. I have no idea how that is with other languages, but I think it is still save to say that either the languages are similar enough to be called the same and don't need translation or are different enough to have their own names.

Despite the similarities of these languages, I'm fine with creating locales like `de-AT.toml`, `de-DE.toml`, `de-CH.toml`, etc, but I'm experiencing technical issues in the linked Hugo Forum thread:
https://discourse.gohugo.io/t/i18n-support-for-language-country-code/20303/4?u=vincenttam.

I'll keep an eye on this because of my native written language: `zh-TW` :hong_kong:, which is quite different from `zh-CN` :cn:.  Dropping support for anyone of these
```toml
defaultContentLanguage = "en"
[languages]
  # Labelled "Chinese" for "Simiplified Chinese", no "Traditional Chinese" support
  [languages.zh]
    weight = 1
    languageCode = "zh-CN"
  [languages.en]
    # …
```
can cause sentiments, :-1: dislikes and complaints from thousands of netizens from these regions in case of big projects like GitLab: https://gitlab.com/gitlab-org/rgitlab-ce/issues/60768.

> I would imagine a menu similar to this:
> 
> * English (UK)
> * English (US)
> * Deutsch (AT)
> * Deutsch (DE)
> * Deutsch (CH)
> * Français (FR)
> * Français (CA)
> * Français (CH)
> 
> And possibly, if there's only one language variation, omit the info in the parenthesis.

@martignoni LGTM, but @<span>pacollins</span> has the final say.

**Comment by pacollins** [comment]
I will defer to @VincentTam to the solution here. As an American, seeing
the English flag to represent English doesn't offend me, so I'd venture
that my opinion doesn't accurately reflect your concerns. My concern is
more the aesthetic of the theme and how we implement whatever we decide.
So @VincentTam, your call. Once we decide what we are doing, I'll comment
from the design side.

**Comment by VincentTam** [comment]
@pacollins @twatzl @martignoni I appreciate the @pacollins' effort in getting the flags displayed on our theme, but @martignoni's suggestion enables us to support more regional languages/dialects.  After spending two days on improving our theme's Germany, French, Polish and Potuguese support in #75, I realized the technical difficulty in handling custom/regional `languageCode`s like `pt-BR`, `de-DE`, etc.  Consider *lingua franca* like Esperanto (which Beautiful Hugo supports) and Latin.  What flag should we use for each of them?  For the later, perhaps :vatican_city: Vatincan's flag. that of Roman Empire :laughing:.

I would further simplify @martignoni's menu into

- English (en)
- Français (fr)
- Deutsch (de)
- Polski (pl)
- Portuguese (pt)
- 正體中文 (zh-TW)
- 简体中文 (zh-CN)

**Comment by pacollins** [comment]
Sounds good. Lets go ahead and begin developing this change? (Volunteers? @twatzl? @martignoni? 😄) Once we get a framework set, I am happy to advice on the css.

1. Switch from flag to text, per @VincentTam.
2. Remove flag resource (css and images) references and files
3. Begin styling new menu (do we want to wait for #46? or just match what we have now?)

**Comment by twatzl** [comment]
Sure. If I find some time I'd be happy to help you. Do we have any means for communication other than this issue thread?

@VincentTam Be careful with CSS dropdowns though. I used that with Hexo, before I ported my blog over to Hugo and they do not work on mobile (because there is no 'hover' event) and I spent a whole day trying to fix it, but could not manage to get it done.

For this reason I would split it into 2 steps. Step 1 change the current menu from flags to languages, Step 2 change the sidebar menu to a dropdown. That way we can still decide to stick with the current solution if changing it causes too much trouble.

**Comment by pacollins** [comment]
> For this reason I would split it into 2 steps. Step 1 change the current menu from flags to languages, Step 2 change the sidebar menu to a dropdown. That way we can still decide to stick with the current solution if changing it causes too much trouble.

I fully agree that there should be separate PRs.

>Do we have any means for communication other than this issue thread?

We prefer to keep it all on GitHub for posterity (in case someone else has the problem). We aren't bothered by double comments or extra commits. That being said, if you really want another means, I can hook you up with my email.

**Comment by VincentTam** [comment]
> Sounds good. Lets go ahead and begin developing this change? (Volunteers? @twatzl? @martignoni? smile) Once we get a framework set, I am happy to advice on the css.
> 
> 1. Switch from flag to text, per @VincentTam.
> 2. Remove flag resource (css and images) references and files
> 3. Begin styling new menu (do we want to wait for #46? or just match what we have now?)

First two steps done in #80.  Click to see the screenshot.  I've never studied CSS, so what you see in the screencap is simply a temporary layout.
